### DISCUSSION TITLE
Cannot run pip as root

### DISCUSSION THREAD

**Discussion by 0Interest** [apache/airflow#33724] [General]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### What happened

Airflow version 2.6.2

I am trying in the Dockerfile to install the package python-lzo, which needs to run gcc in the background, however when in the Dockerfile I switch to root user I get the known bug that occurred in Airflow version 2.3, which to my surprise was already fixed and merged.

I do have the _PIP_ADDITIONAL_REQUIREMENTS: '' set, however it still throws the error.

"You are running pip as root. Please use Airflow user to run pip!"

The Dockerfile is along the lines of:

```
from artifactory.spam/airflow:2.6.2-python3.7-1.0.1
USER root
run pip install python-lzo
USER airflow
```

And I've already tried to set the AIRFLOW_UID=50000 and AIRFLOW_GID=0 and do so in the .env file, although I didn't know where to put it, so I put it in the same folder as the docker-compose.yml and Dockerfile folder.

My PC is Windows 10 however the airflow is running on Linux.

Any ideas?

### What you think should happen instead

It should be able to install python-lzo when running pip as root

### How to reproduce

- Set up an Airflow using docker-compose
- In the Dockerfile try to add USER root and then run pip install afterwards

### Operating System

Debian GNU/Linux 11 (bullseye)

### Versions of Apache Airflow Providers

apache-airflow-providers-common-sql==1.3.2
...-ftp==3.3.0
...-http==4.1.0
...-ssh==3.0.0

The airflow python library is 2.5.0 although the airflow image itself is 2.6.2

### Deployment

Docker-Compose

### Deployment details

I am using docker-compose, although the problem is not really related to it, but more of the Dockerfile

### Anything else

Everytime I run `docker-compose build` I either get the error that I can't run pip as root, or when I remove the `USER root` from the Dockerfile then I get an error that "gcc" has permission error... Because python-lzo library has to be installed using high privilege pip.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by potiuk** [comment]
You should not add PIP packages as root. You should do it as airflow user. Switching to `root` while buiilding the image should only be done for installing `debian` packages. 


See recipes here, they explain everything and show examples of how you can extend airflow image:

https://airflow.apache.org/docs/docker-stack/build.html#quick-start-scenarios-of-image-extending

Of course providing that you are using official airflow image - your "artifactory" image specification does not tell what the base image is.

Converting it to a discussion. In case more discussion is needed.

**Reply by potiuk** [reply]
The docs are pretty clear about it BTW: 

>  When adding packages via pip you need to use the airflow user rather than root. Attempts to install pip packages as root will fail with an appropriate error message.

You have not mentioned the error message, but if you use airflow image it for sure says "don't use root"

**Reply by 0Interest** [reply]
You are correct, I have seen this in the docs, however what if I need  (must) to install a python package which needs elevated permissions such as python-lzo?

**Reply by 0Interest** [reply]
Hi, are there any news? ☺️

**Reply by Taragolis** [reply]
>You are correct, I have seen this in the docs, however what if I need (must) to install a python package which needs elevated permissions such as python-lzo?

Could you give more info here? Which permission it required? 
I thought you need to have gcc, headers for LZO and some additional developer packages. And Jarek @potiuk give you a link with explanation.

BTW I have no issues to install it locally on ARM and by use EOL version of Python and use slim Airflow image

```dockerfile
FROM apache/airflow:slim-2.6.2-python3.7

USER root
RUN apt-get update \
  && apt-get install -y --no-install-recommends build-essential liblzo2-dev

USER airflow
RUN python -m pip install python-lzo
RUN python -m pip list
```

**Reply by 0Interest** [reply]
I am not sure which permission it requires, all I know is that when I do `RUN pip install python-lzo` (or just put it in the `requirements.txt`) it raises an error

```
Error: command gcc failed: Permission denied
```

When I try your answer I get:

Unable to locate liblzo2-dev, we are using a different repository than PyPI, I think we have the .tar.gz of that package though, so how would you install if you have it as .tar.gz?

**Reply by Taragolis** [reply]
[`liblzo2-dev`](https://packages.debian.org/bullseye/liblzo2-dev) it is Debian package, and it should be install thought [`apt-get`](https://linux.die.net/man/8/apt-get)

**Reply by Taragolis** [reply]
Just have a look my sample of Docker image. Without development headers and build-essential you unlikely install `python-lzo`, and changing user can't help here

**Reply by 0Interest** [reply]
Why not just let users use pip with sudo? What is the reasoning? We are the owners of the DAG..

**Reply by 0Interest** [reply]
Also could you give a link to your Dockerfile? @Taragolis

**Reply by Taragolis** [reply]
https://github.com/apache/airflow/discussions/33724#discussioncomment-6823152

```dockerfile
FROM apache/airflow:slim-2.6.2-python3.7

USER root
# Install system packages:
# build-essential - metapakage which include software for compling c and c++ software (Debian)
# liblzo2-dev - data compression library (development files) (Debian)
RUN apt-get update \
  && apt-get install -y --no-install-recommends build-essential liblzo2-dev

USER airflow
RUN python -m pip install python-lzo
RUN python -m pip list
```

**Reply by 0Interest** [reply]
@Taragolis i get "unable go locate liblzo2-dev"

**Reply by Taragolis** [reply]
If you use `artifactory.spam/airflow:2.6.2-python3.7-1.0.1` than it might be:
- Based on non-official image (e.g. non Debian)
- Use custom APT repo with limited packages
- A lot of unknown reason

So if you use custom Airflow image, I would recommend ask first a person who prepare this image why you can't install liblzo2-dev, you also could provide a link to this thread, it might save a time

With official I do not have any issue when I want to install dev packages (by root) and install after that python-lzo (as airflow user), see my output
 
```console
#5 [2/4] RUN apt-get update   && apt-get install -y --no-install-recommends build-essential liblzo2-dev
#5 0.594 Get:1 http://deb.debian.org/debian bullseye InRelease [116 kB]
#5 0.700 Get:2 http://deb.debian.org/debian-security bullseye-security InRelease [48.4 kB]
#5 0.750 Get:3 http://deb.debian.org/debian bullseye-updates InRelease [44.1 kB]
#5 0.884 Get:4 http://deb.debian.org/debian bullseye/main arm64 Packages [8071 kB]
#5 2.405 Get:5 https://packages.microsoft.com/debian/11/prod bullseye InRelease [3629 B]
#5 3.068 Get:6 https://packages.microsoft.com/debian/11/prod bullseye/main arm64 Packages [17.1 kB]
#5 3.105 Get:7 https://apt.postgresql.org/pub/repos/apt bullseye-pgdg InRelease [123 kB]
#5 3.132 Get:8 http://deb.debian.org/debian-security bullseye-security/main arm64 Packages [242 kB]
#5 3.183 Get:9 http://deb.debian.org/debian bullseye-updates/main arm64 Packages [14.7 kB]
#5 3.514 Get:10 https://packages.microsoft.com/debian/11/prod bullseye/main amd64 Packages [101 kB]
#5 5.000 Get:11 https://packages.microsoft.com/debian/11/prod bullseye/main armhf Packages [15.8 kB]
#5 5.000 Get:12 https://packages.microsoft.com/debian/11/prod bullseye/main all Packages [1149 B]
#5 5.368 Get:13 https://apt.postgresql.org/pub/repos/apt bullseye-pgdg/main arm64 Packages [288 kB]
#5 8.211 Fetched 9085 kB in 8s (1155 kB/s)
#5 8.211 Reading package lists...
#5 8.544 Reading package lists...
#5 8.860 Building dependency tree...
#5 8.933 Reading state information...
#5 9.012 The following additional packages will be installed:
#5 9.012   binutils binutils-aarch64-linux-gnu binutils-common bzip2 cpp cpp-10
#5 9.012   dpkg-dev g++ g++-10 gcc gcc-10 libasan6 libatomic1 libbinutils libc-dev-bin
#5 9.012   libc6-dev libcc1-0 libcrypt-dev libctf-nobfd0 libctf0 libdpkg-perl
#5 9.012   libgcc-10-dev libgomp1 libisl23 libitm1 liblsan0 liblzo2-2 libmpc3 libmpfr6
#5 9.013   libnsl-dev libstdc++-10-dev libtirpc-dev libtsan0 libubsan1 linux-libc-dev
#5 9.013   make patch xz-utils
#5 9.013 Suggested packages:
#5 9.013   binutils-doc bzip2-doc cpp-doc gcc-10-locales debian-keyring gcc-10-doc
#5 9.013   gcc-multilib manpages-dev autoconf automake libtool flex bison gdb gcc-doc
#5 9.013   glibc-doc git bzr libstdc++-10-doc make-doc ed diffutils-doc
#5 9.013 Recommended packages:
#5 9.013   fakeroot libalgorithm-merge-perl manpages manpages-dev libc-devtools
#5 9.013   libfile-fcntllock-perl liblocale-gettext-perl
#5 9.126 The following NEW packages will be installed:
#5 9.126   binutils binutils-aarch64-linux-gnu binutils-common build-essential bzip2
#5 9.126   cpp cpp-10 dpkg-dev g++ g++-10 gcc gcc-10 libasan6 libatomic1 libbinutils
#5 9.126   libc-dev-bin libc6-dev libcc1-0 libcrypt-dev libctf-nobfd0 libctf0
#5 9.126   libdpkg-perl libgcc-10-dev libgomp1 libisl23 libitm1 liblsan0 liblzo2-2
#5 9.126   liblzo2-dev libmpc3 libmpfr6 libnsl-dev libstdc++-10-dev libtirpc-dev
#5 9.126   libtsan0 libubsan1 linux-libc-dev make patch xz-utils
#5 9.248 0 upgraded, 40 newly installed, 0 to remove and 9 not upgraded.
#5 9.248 Need to get 56.1 MB of archives.
#5 9.248 After this operation, 210 MB of additional disk space will be used.
...
#6 [3/4] RUN python -m pip install python-lzo
#6 1.719 Collecting python-lzo
#6 1.937   Downloading python-lzo-1.15.tar.gz (17 kB)
#6 1.959   Installing build dependencies: started
#6 5.967   Installing build dependencies: finished with status 'done'
#6 5.969   Getting requirements to build wheel: started
#6 6.192   Getting requirements to build wheel: finished with status 'done'
#6 6.195   Preparing metadata (pyproject.toml): started
#6 6.436   Preparing metadata (pyproject.toml): finished with status 'done'
#6 6.439 Building wheels for collected packages: python-lzo
#6 6.441   Building wheel for python-lzo (pyproject.toml): started
#6 6.870   Building wheel for python-lzo (pyproject.toml): finished with status 'done'
#6 6.871   Created wheel for python-lzo: filename=python_lzo-1.15-cp37-cp37m-linux_aarch64.whl size=27129 sha256=f1a53a4c6b386582c25ff1d5cf42ef2340a2e5879478caa0916694621dcf8143
#6 6.871   Stored in directory: /home/airflow/.cache/pip/wheels/32/97/a2/6dbf942d33ae1036c93a202c422ad8ef3b081c627a33a31e1c
#6 6.875 Successfully built python-lzo
#6 8.770 Installing collected packages: python-lzo
#6 8.787 Successfully installed python-lzo-1.15
#6 8.793 
#6 8.793 [notice] A new release of pip is available: 23.1.2 -> 23.2.1
#6 8.793 [notice] To update, run: python -m pip install --upgrade pip
#6 DONE 8.9s
```

**Reply by potiuk** [reply]
@0Interest -> You jump to conclusions. There is never need to use root to install pip. 

I suggest you to do some search before and read the warnings you get. 

Even when you try to use `pip` as root it even warns you that you should not do it. I suggest to follow helpful advices there.

If you do a simple search in google for the error you mentioned `Error: command gcc failed: Permission denied` you will find out that this is misleading message from `pip` telling you that gcc compiler is missing (if you have a problem with that message being misleading I suggest to raise an issue to `pip`). And it is missing (as our explanation helpfully explains) because `build-essentials` are not installed in order to save 200MB.

Reading our docs is also useful:

https://airflow.apache.org/docs/docker-stack/build.html#extending-the-image 

> Extending the image is easiest if you just need to add some dependencies that do not require compiling. The compilation framework of Linux (so called build-essential) is pretty big, and for the production images, size is really important factor to optimize for, so our Production Image does not contain build-essential. If you need a compiler like gcc or g++ or make/cmake etc. - those are not found in the image

This explains about build-essentials

And you even have example how to install package that needs build-essential https://airflow.apache.org/docs/docker-stack/build.html#example-when-you-add-packages-requiring-compilation

Your package apparently require not only gcc compiler to be installed, but also some libraries that you need to install. As always when you require a package that needs compilation, it's your task (and google search is also helpful with that) to find what packages you need to install to get the package that requires compilation installed. If the package is missing in the Docker image of ours (which is highly likely - you (likely deliberately) chose the slim image as a base which has absolutely minimum set of packages installed needd to run plain airflow core and nothing else - then you need to find out

* whether your package needs compilation (then you need to have build essentials installed)
* which libraries needs to be installed and how

This is always the case - no matter if you are using our docker, another docker or install this package locally. This is just standard way how pip installation is and you need to find out how to install your package.

In the future, I heartily recommend to use google search where you can find a lot of help with "generic" package installation issues - usually when you look for error messages you will find people who installed similar packages and solved similar problems. This is what I usually do and heartily recommend in the future if you have similar problems.

**Reply by potiuk** [reply]
> Why not just let users use pip with sudo? What is the reasoning? We are the owners of the DAG..

Besides the fact that you jumped to conclusions and you did not have problems with sudo or root, answering your question here I heartily recommend Docker documentation and reading about best practices. Using root for docker images as the user to use is a HUGE security risk. You should NEVER do it in production. You can start reading from the official recommendation of Docker https://dockerlabs.collabnix.com/security/Running-Containers-as-ROOT.html - but if you look at it they have helpful links and resources that you can follow that explain all the details. I hope I answered your question.

Also since you are apparently interested in understanding some of the design decisions, I also recommend you to read this part of our documentation https://airflow.apache.org/docs/docker-stack/entrypoint.html#allowing-arbitrary-user-to-run-the-container  that explains all details about how our image should be run following best practices - including OpenShift requirements. It's a very interesting read and has links to resources.

**Reply by 0Interest** [reply]
@potiuk Hello, firstly I would like to apologize if my messages were took as offensive or it was looked like I am mad, I am not.

I did not jump to conclusions, I just honestly ask how come you state that 

1. You cannot run pip as root because it is super dangerous

However..

2. It is possible to run pip as root, and the message I get is no more than a warning that `build-essential` is not installed

Does that mean that if I install `build-essential` I would be able to install it using root? The message of the error will still show however it would still work because it is not an error but a warning?

Also, why would I need more than gcc to install python-lzo? If pip install python-lzo gets all the packages from the artifactory, why would airflow Dockerfile need to do something other than just the same command? "RUN pip install python-lzo"?

Thanks

**Reply by potiuk** [reply]
* You SHOULD NOT run pip as root (and pip tell you so when you try)
* You SHOULD NOT run the image as root (and the documantation says so - please read it).

The fact that you CAN do something does not mean you SHOULD do it

> Does that mean that if I install build-essential I would be able to install it using root? The message of the error will still show however it would still work because it is not an error but a warning?

You CAN do what you want. But in our image it won't work because of the above things and our opinionated decission to ask our users to instal all pip packages as airflow user. And if you run something as root, you are breaking the assumptions and opinionated decision we made (and clearly documented and even went to a great lenght to replace pip in the image to tell you explicitly to NOT do it). So no - it won't work with our image. 

You migt follow our opinionsand use our image, or follow your own insticts and build your own image from scratch in the way that you think is better. If you think you shoudl be able to have your image with root installation - feel free. But it will not work with Airflow image by the community, because we specififcally made it not work with root - to protect our users. If you disagree with this decision - feel free, you can use your own image.

> Also, why would I need more than gcc to install python-lzo? If pip install python-lzo gets all the packages from the artifactory, why would airflow Dockerfile need to do something other than just the same command? "RUN pip install python-lzo"?

No. Pip does not take packages from artifactory. It downloads them from `pypi`. And no your assumption is wrong. When python package needs compilation, it might need additional libraries when there is no binary wheel available in `pypi` matching your architecture. The package of your choice `python-lzo` apparently does not have binary wheel, so pip needs to compile some libraries and this means it needs to have some dependencies that are installed in the system - not dependent on what is available in PyPI. This is how - currenlty - PyPI ecosystem works. Ther are  a few PEPs' Python Improvement Proposals being discussed /implemented in Python community https://peps.python.org/topic/packaging/ that might change it in the future, but it's many years until they might be implemented an deployed.

If you have a package that has some dependencies it's your task to find out what they are and make sure they are installed. This is how it works.

**Reply by 0Interest** [reply]
@Taragolis 
When I try to apt update I get:

"The repository deb.debian.org does not have a release file"

How do I change it so the sources.list have the correct path?

**Reply by 0Interest** [reply]
@potiuk is it possible to run airflow image on ubuntu? We do not have a repository for debian in our network, only ubuntu. Changing the `sources.list` to the ubuntu server does not help either, I get some authentication error even though I set up [trusted=true], added the flag --allow-unauthenticated and such to the apt update command... Nothing works.

the ubuntu repo im using is
Ubuntu focal

**Reply by potiuk** [reply]
You can build your own image from scratch. Our image is using debian.
